File: packages/extractor/src/insightgraph_extractor/pipeline.py
from __future__ import annotations

import asyncio
import logging

from insightgraph_core.ir.extraction import (
    ExtractedClaim,
    ExtractedEntity,
    ExtractedMetric,
    ExtractionResult,
)
from insightgraph_core.ir.models import Block, DocumentIR
from insightgraph_core.types import BlockType
from insightgraph_extractor.claim import ClaimExtractor
from insightgraph_extractor.entity import EntityExtractor
from insightgraph_extractor.metric import MetricExtractor

logger = logging.getLogger(__name__)

# Block types that contain natural-language text suitable for entity / claim
# extraction.
_TEXT_BLOCK_TYPES = {BlockType.PARAGRAPH, BlockType.HEADING}


class ExtractionPipeline:
    """Orchestrates entity, metric, and claim extraction over a DocumentIR.

    All three extractors run concurrently via ``asyncio.gather``.
    """

    def __init__(
        self,
        model: str = "gpt-4o-mini",
        api_key: str = "",
    ) -> None:
        self.entity_extractor = EntityExtractor(model=model, api_key=api_key)
        self.metric_extractor = MetricExtractor(model=model, api_key=api_key)
        self.claim_extractor = ClaimExtractor(model=model, api_key=api_key)

    async def extract(self, doc: DocumentIR) -> ExtractionResult:
        """Run the full extraction pipeline on a parsed document.

        Args:
            doc: The intermediate representation of the parsed document.

        Returns:
            An ``ExtractionResult`` containing all extracted entities, metrics,
            and claims.
        """
        all_blocks: list[Block] = list(doc.iter_all_blocks())
        text_blocks: list[Block] = [b for b in all_blocks if b.type in _TEXT_BLOCK_TYPES]

        context = {"title": doc.title or doc.source_filename}

        logger.info(
            "Starting extraction pipeline for document %s (%d total blocks, %d text blocks)",
            doc.id,
            len(all_blocks),
            len(text_blocks),
        )

        results = await asyncio.gather(
            self.entity_extractor.extract(text_blocks, context),
            self.metric_extractor.extract(all_blocks, context),
            self.claim_extractor.extract(text_blocks, context),
            return_exceptions=True,
        )

        entities: list[ExtractedEntity] = []
        metrics: list[ExtractedMetric] = []
        claims: list[ExtractedClaim] = []

        for idx, result in enumerate(results):
            if isinstance(result, BaseException):
                extractor_name = ("entity", "metric", "claim")[idx]
                logger.error("%s extractor failed: %s", extractor_name, result)
                continue
            if idx == 0:
                entities = result
            elif idx == 1:
                metrics = result
            else:
                claims = result

        logger.info(
            "Extraction complete for document %s: %d entities, %d metrics, %d claims",
            doc.id,
            len(entities),
            len(metrics),
            len(claims),
        )

        return ExtractionResult(
            document_id=doc.id,
            entities=entities,
            metrics=metrics,
            claims=claims,
        )

File: packages/extractor/src/pipeline.ts
import type {
  DocumentIR,
  Block,
  ExtractionResult,
  DomainConfig,
  BlockType,
} from "@insightgraph/core";
import { iterAllBlocks } from "@insightgraph/core";
import { EntityExtractor } from "./entity";
import { MetricExtractor } from "./metric";
import { ClaimExtractor } from "./claim";
import { RelationshipExtractor } from "./relationship";
import type { ExtractorOptions } from "./options";

const DATA_BLOCK_TYPES: Set<BlockType> = new Set(["paragraph", "heading", "data_row"]);

export class ExtractionPipeline {
  private entityExtractor: EntityExtractor;
  private metricExtractor: MetricExtractor;
  private claimExtractor: ClaimExtractor;
  private relationshipExtractor: RelationshipExtractor;
  private domain?: DomainConfig;

  constructor(
    model: string,
    apiKey: string,
    baseUrl = "",
    domainConfig?: DomainConfig,
    options: ExtractorOptions = {},
  ) {
    this.entityExtractor = new EntityExtractor(model, apiKey, baseUrl, options);
    this.metricExtractor = new MetricExtractor(model, apiKey, baseUrl, options);
    this.claimExtractor = new ClaimExtractor(model, apiKey, baseUrl, options);
    this.relationshipExtractor = new RelationshipExtractor(model, apiKey, baseUrl, options);
    this.domain = domainConfig;
  }

  /**
   * Run the four extraction stages. Entities, metrics, and claims are all
   * independent — they run *concurrently* via Promise.all. The relationship
   * extractor needs the list of entity names for its prompt and post-filter,
   * so it runs in a second wave.
   *
   * Before parallelization, a large document with ~2000 blocks × 4 categories
   * × concurrency=4 took ~20 minutes (effective concurrency of 4 across the
   * whole pipeline). Running 3 categories in parallel triples the effective
   * concurrency to ~12 during phase 1 and cuts total extraction time roughly
   * in half for typical documents.
   */
  async extract(doc: DocumentIR): Promise<ExtractionResult> {
    const allBlocks = Array.from(iterAllBlocks(doc));
    const textBlocks: Block[] = allBlocks.filter((b) => DATA_BLOCK_TYPES.has(b.type));

    const context: { title: string; domain?: DomainConfig } = {
      title: doc.title ?? doc.sourceFilename,
      domain: this.domain,
    };

    console.log(`[pipeline] Starting extraction: ${textBlocks.length} blocks`);

    // Phase 1: entities + metrics + claims run in parallel. Each extractor
    // manages its own concurrency internally; the outer Promise.allSettled
    // lets their LLM calls interleave so idle RTT on one category is
    // backfilled by work on the others, and — critically — a failure in one
    // category does NOT discard the successful output of the others.
    const phase1Start = Date.now();
    const [entitiesRes, metricsRes, claimsRes] = await Promise.allSettled([
      this.entityExtractor.extract(textBlocks, context),
      this.metricExtractor.extract(textBlocks, context),
      this.claimExtractor.extract(textBlocks, context),
    ]);
    const pickOr = <T>(
      res: PromiseSettledResult<T[]>,
      label: string,
    ): T[] => {
      if (res.status === "fulfilled") return res.value;
      console.warn(
        `[pipeline] ${label} extraction rejected: ${(res.reason as Error)?.message ?? res.reason}`,
      );
      return [];
    };
    const entities = pickOr(entitiesRes, "entity");
    const metrics = pickOr(metricsRes, "metric");
    const claims = pickOr(claimsRes, "claim");
    console.log(
      `[pipeline] Phase 1 done in ${Date.now() - phase1Start}ms: ` +
        `${entities.length} entities, ${metrics.length} metrics, ${claims.length} claims`,
    );

    // Phase 2: relationships depend on phase 1 entity names (used both in
    // the prompt and to post-filter hallucinated endpoints).
    const phase2Start = Date.now();
    const entityNames = entities.map((e) => e.name);
    const relationships = await this.relationshipExtractor.extract(textBlocks, {
      title: context.title,
      entityNames,
    });
    console.log(
      `[pipeline] Phase 2 done in ${Date.now() - phase2Start}ms: ` +
        `${relationships.length} relationships`,
    );

    return {
      documentId: doc.id,
      entities,
      metrics,
      claims,
      relationships,
      resolvedEntities: [],
    };
  }
}
